src/twisted/internet/test/test_asyncfs.py

```diff
@@ -0,0 +1,341 @@
+        return d
+
+    def test_makelink(self):
+        def cb_link(_):
+            with open("one.lnk") as fd:
+                self.assertEqual(fd.read(), "blahblahblah")
+
+        d = self.fso.makeLink("one.lnk", "one.txt")
+        d.addCallback(cb_link)
+        return d
+
+    def test_readlink(self):
+        def cb_link2(r):
+            self.assertEqual(r, os.path.join(self.tpath, "one.txt"))
+
+        def cb_link1(_):
+            d2 = self.fso.readLink("one.lnk")
+            d2.addCallback(cb_link2)
+            return d2
+
+        if platform.system() == "Windows":
+            raise unittest.SkipTest("doesn't work on Windows")
+        d1 = self.fso.makeLink("one.lnk", "one.txt")
+        d1.addCallback(cb_link1)
+        return d1
+
+    def test_listdir(self):
+        def cb_listdir(r):
+            r = list(r)
+            self.assertDir([name for name, attrs in r])
+            r = [attrs for name, attrs in r if name == "one.txt"]
+            self.assertEqual(r[0]["mtime"], 4000)
+
+        os.utime("one.txt", (4000, 4000))
+        d1 = self.fso.openDirectory(".")
+        d1.addCallback(cb_listdir)
+        return d1
+
+    def test_statfs(self):
+        def cb_statfs(r):
+            log.debug("statvfs: {vfs}", vfs=r)
+            v = os.statvfs(".")
+            self.assertEqual(r["size"], v.f_frsize)
+            self.assertEqual(r["blocks"], v.f_blocks)
+            self.assertTrue(abs(r["free"] - v.f_bavail) < 5)  # free blocks can change
+
+        if not hasattr(os, "statvfs"):
+            raise unittest.SkipTest("no statvfs on this platform")
+        d1 = self.fso.statfs()
+        d1.addCallback(cb_statfs)
+        return d1
+
+    def test_consumer(self):
+        def cb_con2(_, fd, consumer):
+            self.assertEqual(consumer.total_bytes, len(ONE_MEG_OF_BLAH))
+            return fd.close()
+
+        def cb_con1(fd):
+            consumer = FakeConsumer()
+            consumer.registerProducer(fd.producer, True)
+            d2 = fd.send(consumer)
+            d2.addCallback(cb_con2, fd, consumer)
+            return d2
+
+        with open("bigfile.txt", "w") as fd:
+            fd.write(ONE_MEG_OF_BLAH)
+        d1 = self.fso.openFile("bigfile.txt", os.O_CREAT | os.O_RDONLY)
+        d1.addCallback(cb_con1)
+        return d1
+
+    def test_producer(self):
+        def cb_pro2(_):
+            s = os.stat("bigfile.txt")
+            self.assertEqual(s.st_size, len(ONE_MEG_OF_BLAH))
+
+        def cb_pro1(fd):
+            producer = FakeProducer()
+            consumer = fd.receive()
+            consumer.registerProducer(producer, True)
+            producer.fire(consumer)
+            d2 = fd.close()
+            d2.addCallback(cb_pro2)
+            return d2
+
+        d1 = self.fso.openFile("bigfile.txt", os.O_CREAT | os.O_WRONLY)
+        d1.addCallback(cb_pro1)
+        return d1
+
+
+class TestThreadFs(BaseTestFs, unittest.TestCase):
+    fso_class = asyncfs.ThreadFs
```

Comment:
Mixins are an anti-pattern.  For this kind of test sharing (speculative at this point, I guess), composition is preferable.  I know unittest doesn't make it easy but it's still better than dealing with the complexity of mixins.

Also, I'd suggest that stateful Hypothesis tests might just make this test suite shorter, easier to read, and more comprehensive.